firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -180,44 +177,51 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
   @NonNull
   public UpdateTask updateApp() throws FirebaseAppDistributionException {
 
-    if (updateAppTaskCompletionSource != null
-        && !updateAppTaskCompletionSource.getTask().isComplete()) {
-      updateAppCancellationSource.cancel();
-    }
-
-    updateAppCancellationSource = new CancellationTokenSource();
-    updateAppTaskCompletionSource =
-        new TaskCompletionSource<>(updateAppCancellationSource.getToken());
-    Context context = firebaseApp.getApplicationContext();
-    this.updateTask = new UpdateTaskImpl(updateAppTaskCompletionSource.getTask());
+    if (isTesterSignedIn()) {
+      if (updateAppTaskCompletionSource != null
+          && !updateAppTaskCompletionSource.getTask().isComplete()) {
+        updateAppCancellationSource.cancel();
+      }
 
-    if (cachedLatestRelease == null) {
-      setUpdateAppTaskCompletionError(
-          new FirebaseAppDistributionException(
-              context.getString(R.string.no_update_available),
-              FirebaseAppDistributionException.Status.UPDATE_NOT_AVAILABLE));
-    } else {
-      if (cachedLatestRelease.getBinaryType() == BinaryType.AAB) {
-        redirectToPlayForAabUpdate(cachedLatestRelease.getDownloadUrl());
+      updateAppCancellationSource = new CancellationTokenSource();
+      updateAppTaskCompletionSource =
+          new TaskCompletionSource<>(updateAppCancellationSource.getToken());
+      Context context = firebaseApp.getApplicationContext();
+      this.updateTask = new UpdateTaskImpl(updateAppTaskCompletionSource.getTask());
+
+      if (cachedLatestRelease == null) {
```

Comment:
Can we move this check earlier in the function? Then we can return early if the cachedLatestRelease is null, create a UpdateTaskImpl error with Tasks.forException, and save the overhead of creating a completion sources for a task that will fail

Reply:
If you move the sign in check early, the flow could be:
 if (!isTesterSignedIn(){
// ...
}
if (cachedLatestRelease == null) { 
// ...
}
// else tester is signed in and cachedLatestRelease is non-null
....